File: py/03-retos-programacion/ejercicio-00-sintaxis-hola-mundo.py
# EJERCICIO 00-SINTAXIS, VARIABLES, TIPOS DE DATOS Y HOLA MUNDO

# Documentación oficial del lenguaje Python: https://www.python.org/doc/
# Aunque python.org es la página principal, la sección de 'doc' es donde se encuentra la documentación detallada.

# Esto es un comentario de una sola línea en Python.
# Puedes usarlo para añadir notas rápidas o deshabilitar temporalmente una línea de código.

"""
Esto es un comentario de varias líneas (docstring) utilizando comillas dobles.
Generalmente, los 'docstrings' se usan para documentar módulos, funciones,
clases y métodos. Se pueden acceder en tiempo de ejecución.
"""

'''
Esto es otro ejemplo de comentario de varias líneas (docstring) utilizando comillas simples.
Funciona de la misma manera que las comillas dobles para este propósito.
La convención principal es usar comillas dobles para 'docstrings'.
'''

# Variables en Python
# Python es un lenguaje de tipado dinámico, lo que significa que no necesitas
# declarar explícitamente el tipo de una variable. El intérprete lo infiere.
my_variable = "Soy una variable"
my_other_variable = 10 # Las variables pueden cambiar su tipo durante la ejecución.

# Constantes en Python
# Python no tiene un concepto de "constante" estricto como otros lenguajes.
# Por convención, se usan mayúsculas para indicar que una variable debe
# ser tratada como una constante y no modificarse.
MY_CONSTANT = "Soy una constante (por convención)"

# Tipos de datos primitivos (Built-in Types)
# Estos son los tipos fundamentales que Python soporta de forma nativa.

# Tipo de dato 'str' (cadena de texto)
my_string = "Soy un string"
my_other_string = 'También soy un string' # Las comillas simples o dobles son válidas.

# Tipo de dato 'int' (entero)
my_int = 1
my_negative_int = -5

# Tipo de dato 'float' (número de punto flotante)
my_float = 1.0
my_scientific_float = 1.2e-3 # Notación científica

# Tipo de dato 'bool' (booleano)
# Representa valores de verdad: Verdadero o Falso.
my_bool_true = True
my_bool_false = False

# Tipo de dato 'NoneType' (Ninguno)
# Representa la ausencia de valor. Es similar a 'null' en otros lenguajes.
my_none_value = None

# Números enteros grandes (arbitrary-precision integers)
# Python maneja enteros de precisión arbitraria, lo que significa que
# pueden ser tan grandes como la memoria del sistema lo permita.
my_big_number = 1234567890123456789012345678901234567890

# Imprimir mensajes y variables en la consola
# La función 'print()' se usa para mostrar información en la salida estándar.
print("¡Hola, Python!")

# Concatenación de cadenas y uso de f-strings (una forma moderna y recomendada)
print("Mi variable es: " + my_variable) # Concatenación básica
print(f"Mi variable es: {my_variable}") # Uso de f-string (Python 3.6+)

# Imprimir el tipo de dato de las variables
# La función 'type()' devuelve el tipo de un objeto.
print(f"El tipo de 'my_string' es: {type(my_string)}")
print(f"El tipo de 'my_int' es: {type(my_int)}")
print(f"El tipo de 'my_float' es: {type(my_float)}")
print(f"El tipo de 'my_bool_true' es: {type(my_bool_true)}")
print(f"El tipo de 'my_none_value' es: {type(my_none_value)}")
print(f"El tipo de 'my_big_number' es: {type(my_big_number)}")

# Operaciones básicas con variables
print(f"Suma de enteros: {my_int + my_other_variable}")
print(f"Multiplicación de flotantes: {my_float * my_scientific_float}")
print(f"Negación de un booleano: {not my_bool_true}")

File: js/03-retos-programacion/ejercicio-00-sintaxis-hola-mundo.js
// EJERCICIO 00-SINTAXIS, VARIABLES, TIPOS DE DATOS Y HOLA MUNDO

// Web oficial (documentación) del lenguaje: https://developer.mozilla.org/es/docs/Web/JavaScript
// MDN Web Docs es la fuente de documentación más completa y confiable para JavaScript.

// Este es un comentario de una sola línea en JavaScript.
// Se usa para añadir notas o deshabilitar código temporalmente.

/**
 * Este es un comentario de varias líneas en JavaScript.
 * Es útil para bloques de texto más extensos, como la descripción
 * de funciones o archivos completos.
 */

// Variables en JavaScript
// JavaScript es un lenguaje de tipado dinámico, no necesitas declarar
// explícitamente el tipo de dato de una variable.
let myVariable = "Soy una variable con let";
// 'let' permite reasignar el valor y tiene un ámbito de bloque.

var myOldVariable = 10;
// 'var' es una forma más antigua de declarar variables.
// Tiene un ámbito de función o global y puede causar 'hoisting'.
// Se desaconseja su uso en código moderno a favor de 'let' y 'const'.

const MY_CONSTANT = "Soy una constante con const";
// 'const' declara una constante cuyo valor no puede ser reasignado.
// Su ámbito es de bloque y es ideal para valores que no cambiarán.

// Tipos de datos primitivos (Primitive Data Types)
// JavaScript tiene varios tipos de datos primitivos incorporados.

// Tipo de dato 'string' (cadena de texto)
let myString = "Soy un string con comillas dobles";
let myOtherString = "También soy un string con comillas simples";
let myTemplateString = `Puedo ${myVariable} y expresiones, soy un template string.`;
// Los template strings (backticks ``) permiten incrustar expresiones `${}`
// y manejar saltos de línea sin caracteres especiales.

// Tipo de dato 'number' (número entero o de punto flotante)
let myInt = 1;
let myFloat = 1.0;
let myNegativeNumber = -5;
let myScientificNumber = 1.2e-3; // Notación científica
// JavaScript no diferencia entre enteros y flotantes a nivel de tipo;
// ambos son 'number'.

// Tipo de dato 'boolean' (booleano)
// Representa valores de verdad: true o false.
let myBooleanTrue = true;
let myBooleanFalse = false;

// Tipo de dato 'null'
// Representa la ausencia intencional de cualquier valor o un valor desconocido.
let myNullValue = null;

// Tipo de dato 'undefined'
// Indica que una variable ha sido declarada, pero aún no se le ha asignado un valor.
let myUndefinedValue; // Automáticamente se inicializa como undefined.

// Tipo de dato 'symbol' (símbolo) - Introducido en ES6
// Los símbolos son valores únicos e inmutables que se pueden usar
// como identificadores de propiedades de objetos.
let mySymbol = Symbol("Descripción del símbolo");

// Tipo de dato 'BigInt' (entero de precisión arbitraria) - Introducido en ES2020
// Permite representar números enteros mucho más grandes que los que 'number' puede manejar.
let myBigInt = 1234567890123456789012345678901234567890n;
// Se define añadiendo 'n' al final del número.

// Imprimir mensajes y variables en la consola
// La función 'console.log()' se usa para mostrar información en la consola del navegador
// o en un entorno de ejecución de JavaScript como Node.js.
console.log("¡Hola, JavaScript!");

// Concatenación de cadenas y uso de template strings (la forma moderna y recomendada)
console.log("Mi variable con let es: " + myVariable); // Concatenación básica con '+'
console.log(`Mi constante es: ${MY_CONSTANT}`); // Uso de template string para interpolación
console.log(`Mi variable var es: ${myOldVariable}`);

// Imprimir el tipo de dato de las variables
// El operador 'typeof' devuelve una cadena indicando el tipo del operando.
console.log(`El tipo de 'myString' es: ${typeof myString}`);
console.log(`El tipo de 'myInt' es: ${typeof myInt}`);
console.log(`El tipo de 'myFloat' es: ${typeof myFloat}`);
console.log(`El tipo de 'myBooleanTrue' es: ${typeof myBooleanTrue}`);
console.log(`El tipo de 'myNullValue' es: ${typeof myNullValue}`); // ¡Ojo! typeof null es 'object' (considerado un bug histórico)
console.log(`El tipo de 'myUndefinedValue' es: ${typeof myUndefinedValue}`);
console.log(`El tipo de 'mySymbol' es: ${typeof mySymbol}`);
console.log(`El tipo de 'myBigInt' es: ${typeof myBigInt}`);

// Operaciones básicas con variables
console.log(`Suma de números: ${myInt + myOldVariable}`);
console.log(`Multiplicación de flotantes: ${myFloat * myScientificNumber}`);
console.log(`Negación de un booleano: ${!myBooleanTrue}`); // El operador '!' niega un valor booleano.
